feat(types): Introduce SafePromiseLike (#5506)
fastify/fastify: docs/Reference/TypeScript.md

- Make sure the `no-unused-vars` rule is enabled in
  [ESLint](https://eslint.org/docs/rules/no-unused-vars) and any imported plugin
  are actually being loaded.
- In case you've the `@typescript-eslint/no-floating-promises` enabled,
please double-check that your `.eslintrc` includes `allowForKnownSafePromises`
property as described [here](https://typescript-eslint.io/rules/no-floating-promises/#allowforknownsafepromises):
```
{
  "rules": {
    "@typescript-eslint/no-floating-promises": ["error", {
      "allowForKnownSafePromises": [

>         { "from": "package", "name": "SafePromiseLike", "package": "fastify" }

Reviewer: I set up a reproduction here: https://github.com/JoshuaKGoldberg/repros/tree/fastify-no-floating-promises-repros

...but I couldn't find any cases where `PromiseLike` types are exposed to users. Out of the two issues with repro cases I could find from [`is:issue no-floating-promises`](https://github.com/fastify/fastify/issues?q=is%3Aissue+no-floating-promises):
* #4246: triggers on `reply.header()`, which returns `FastifyReply`
* #4399: triggers on `reply.send()`, which returns `FastifyReply`

In fact, there's no reference to `PromiseLike` in `fastify/types/reply.d.ts` at all. Most methods return `FastifyReply`.

So, can we get away with just recommending using `allowForKnownSafePromises` on `FastifyReply`, and not making any `.d.ts` changes?

```suggestion
        { "from": "package", "name": "FastifyReply", "package": "fastify" }
```

(or: are there cases where `PromiseLike` is indeed exposed to users?)

Reviewer 2: `PromiseLike` is exposed via https://github.com/fastify/fastify/blob/c33cc90802b46570139430999cf290b85feb1856/types/instance.d.ts#L173. We use that technique instead of declaring a `.then()` method in `FastifyInstance` because typescript-eslint would then require the users to add `await` everywhere.

Something that this is not addressing is:
 https://github.com/fastify/fastify/blob/c33cc90802b46570139430999cf290b85feb1856/types/reply.d.ts#L77 and

Reviewer: In the meantime:

```suggestion
        { "from": "package", "name": "FastifyReply", "package": "fastify" },
        { "from": "package", "name": "SafePromiseLike", "package": "fastify" },
```

Author: @JoshuaKGoldberg  should we also add `FastifyInstance` to the list?

Reviewer: I don't see a `.then()` for `FastifyInstance`, do you have a repro case showing it triggering `no-floating-promises`?

Author: Yes, in this case the behavior it's extremely weird.
```
fastify(); // cause the error no-floating-promises
const something = fastify(); // no error
```
So assigning `fastify` to a variable, avoid the `eslint` error 🤔

Reviewer: Now I'm curious about the API design: why would one not assign `fastify()` / a `FastifyInstance` somewhere? And/or, why does it have a `.then()`?

Author: @mcollina know for sure :)

Reviewer 2: > And/or, why does it have a .then()?

It does have a `then()` so that it start loading all plugins. We are not typing it because otherwise _every single method_ in Fastify would need to be awaited according to typescript-eslint and users have been complaining about that. In other terms, a user can do `await app.register(myPlugin)` instead of `app.register(myPlugin); await app.ready()`. It's syntactic sugar that improves the developer experience.

Reviewer: > In other terms, a user can do `await app.register(myPlugin)` instead of `app.register(myPlugin); await app.ready()`.

Right, thanks, this is helpful - I hadn't seen folks doing this (in part because I don't often work in the areas of code Fastify is made for). I'll mention in https://github.com/typescript-eslint/typescript-eslint/issues/8433 too.

- Make sure the `no-unused-vars` rule is enabled in
  [ESLint](https://eslint.org/docs/rules/no-unused-vars) and any imported plugin
  are actually being loaded.
- In case you've the `@typescript-eslint/no-floating-promises` enabled,
please double-check that your `.eslintrc` includes `allowForKnownSafePromises`

> property as described [here](https://github.com/fastify/fastify/issues/5498):

Reviewer 2: Can we point to the official docs instead?

Author: Done on [52c9dab](https://github.com/fastify/fastify/pull/5506/commits/52c9dab1eb4537752aa882acebbbb8187c2c6c9b)

```
{
  "rules": {
    "@typescript-eslint/no-floating-promises": ["error", {
      "allowForKnownSafePromises": [

>         // { "from": "package", "name": "FastifySafePromiseOrSomeSuch", "package": "@fastify/etc" },

Reviewer 2: Can this be removed? Alternatively, can you specify an actual plugin/real case?

Author: Removed on [52c9dab](https://github.com/fastify/fastify/pull/5506/commits/52c9dab1eb4537752aa882acebbbb8187c2c6c9b)

>         { "from": "file", "name": "SafePromise"}

Reviewer 2: Does this capture all of Fastify?

Author: Updated on [52c9dab](https://github.com/fastify/fastify/pull/5506/commits/52c9dab1eb4537752aa882acebbbb8187c2c6c9b)

- Make sure the `no-unused-vars` rule is enabled in
  [ESLint](https://eslint.org/docs/rules/no-unused-vars) and any imported plugin
  are actually being loaded.
- In case you've the `@typescript-eslint/no-floating-promises` enabled,

> please double-check that your `.eslintrc` includes `allowForKnownSafePromises`

Reviewer: Now that [ESLint's new flat config](https://eslint.org/blog/2022/08/new-config-system-part-2) is the default, `.eslintrc` is no longer what many repos will use instead of `eslint.config.*`:

```suggestion
please double-check that your ESLint configuration includes a `allowForKnownSafePromises`
```

property as described on the [`typescript-eslint no-floating-promises allowForKnownSafePromises
documentation`](https://typescript-eslint.io/rules/no-floating-promises/#allowforknownsafepromises):
```
{
  "rules": {
    "@typescript-eslint/no-floating-promises": ["error", {
      "allowForKnownSafePromises": [

>         { "from": "package", "name": "FastifyReply", "package": "fastify" },

Reviewer: Per https://github.com/fastify/fastify/pull/5506#discussion_r1631165436:

```suggestion
        { "from": "package", "name": "FastifyInstance", "package": "fastify" },
        { "from": "package", "name": "FastifyReply", "package": "fastify" },
```
